File: lib/views/sensors_list_page.dart
import 'package:flutter/material.dart';
import 'sensor_details_page.dart';

/// A page that displays a list of all IoT sensors.
///
/// This page allows users to view the sensors and navigate to their
/// details page for further interactions such as updating or deleting
/// sensor data.
class SensorsListPage extends StatelessWidget {
  final List<Map<String, dynamic>> sensorsData; // List of sensor data
  final Function onDataChanged;  // Callback for notifying data changes

  SensorsListPage({required this.sensorsData, required this.onDataChanged});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('All IoT Sensors'), // Title of the app bar
      ),
      body: ListView.builder(
        itemCount: sensorsData.length, // Number of sensors to display
        itemBuilder: (context, index) {
          final sensor = sensorsData[index]; // Retrieve the sensor data for this index
          return ListTile(
            leading: Icon(Icons.sensors), // Icon for the sensor
            title: Text(sensor['sensor_id']), // Display the sensor ID
            subtitle: Text('Category: ${sensor['category']}'), // Display the sensor category
            trailing: Icon(Icons.arrow_forward), // Arrow icon indicating navigation
            onTap: () {
              // Navigate to the SensorDetailsPage when tapped
              Navigator.push(
                context,
                MaterialPageRoute(
                  builder: (context) => SensorDetailsPage(
                    sensorData: sensor, // Pass the selected sensor data
                    onUpdate: (updatedSensor) {
                      // Update the sensor data in the list
                      sensorsData[index] = updatedSensor; // Update the sensor in the list
                      onDataChanged();  // Call the data change callback to save the data
                    },
                    onDelete: () {
                      // Remove the sensor from the list
                      sensorsData.removeAt(index); // Remove the sensor from the list
                      onDataChanged();  // Save the updated data after deletion
                      Navigator.pop(context); // Navigate back to the previous page
                    },
                  ),
                ),
              );
            },
          );
        },
      ),
    );
  }
}

File: frontend/src/components/SensorsListPage.js
import React from 'react';
import { useNavigate } from 'react-router-dom';
import SensorDetailsPage from './SensorDetailsPage'; // Assumes SensorDetailsPage is defined similarly

function SensorsListPage({ sensorsData, onDataChanged }) {
    const navigate = useNavigate();

    // Handler when a sensor item is clicked
    const handleSensorClick = (sensor, index) => {
        // Navigate to the sensor details page, passing sensor data and callbacks via state
        navigate('/sensor-details', {
            state: {
                sensorData: sensor,
                sensorIndex: index,
                onUpdate: (updatedSensor) => {
                    // Create a new sensors array with the updated sensor
                    const newSensorsData = sensorsData.map((s, i) =>
                        i === index ? updatedSensor : s
                    );
                    onDataChanged(newSensorsData);
                },
                onDelete: () => {
                    // Create a new sensors array without the deleted sensor
                    const newSensorsData = sensorsData.filter((_, i) => i !== index);
                    onDataChanged(newSensorsData);
                    navigate(-1); // Navigate back to the previous page
                },
            },
        });
    };

    return (
        <div>
            {/* App Bar */}
            <header style={{ backgroundColor: '#6200EE', color: '#FFF', padding: '16px' }}>
                <h1>All IoT Sensors</h1>
            </header>

            {/* List of Sensors */}
            <ul style={{ listStyleType: 'none', padding: 0 }}>
                {sensorsData.map((sensor, index) => (
                    <li
                        key={sensor.sensor_id}
                        onClick={() => handleSensorClick(sensor, index)}
                        style={{
                            display: 'flex',
                            alignItems: 'center',
                            padding: '16px',
                            borderBottom: '1px solid #ccc',
                            cursor: 'pointer',
                        }}
                    >
                        {/* Leading icon (using an emoji as a placeholder for Icons.sensors) */}
                        <span style={{ marginRight: '16px' }}>🖧</span>
                        <div style={{ flex: 1 }}>
                            {/* Sensor ID */}
                            <div>{sensor.sensor_id}</div>
                            {/* Sensor category */}
                            <div>Category: {sensor.category}</div>
                        </div>
                        {/* Trailing arrow icon (using an emoji as a placeholder) */}
                        <span>➡️</span>
                    </li>
                ))}
            </ul>
        </div>
    );
}

export default SensorsListPage;
